In this project, we work with a dataset of used cars from eBay Kleinanzeigen, a classified section of the Geramn ebay website. The aim of this project to perform data cleaning on the selected data set, then study and analyze the data

# Read in the data

In [1]:
import numpy as np
import pandas as pd
import csv 
autos = pd.read_csv("autos.csv", encoding = "Latin-1")

In [2]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

In the above cells, we read autos.csv file into pandas Dataframe named autos. 
The dataframe consists of 50000 rows and 20 columns
The index or row axis is integer labeled and the column axis is string labeled. 
In the next step we acquired the information on columns of the Dataframe autos. We find that 5 columns - year of registeration, month of registration, power ps, no of pictures, postal code are of integer datatype and rest of them are stored as strings datatype. It can be seen that many of the columns have no null values. Few of the columns have null values but no columns have more than ~20% null values

# Cleaning the column names

In [3]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

We clean the column names in two steps 
1. we replace some of the column labels with names more readable and appropriate based on the data dictionary to be more descriptive 
2. convert the rest of the column labels from camel case to snake case 

In [4]:
#Function to convert camel case to snake case
def change_case(str): 
      
    return ''.join(['_'+i.lower() if i.isupper()  
               else i for i in str]).lstrip('_') 

#Extracting columns labels from autos and storing them in cols array
cols = autos.columns
new_cols = []
for c in cols:
    if c == "yearOfRegistration":
        new_c = "registration_year"
        new_cols.append(new_c)
    elif c == "monthOfRegistration":
        new_c = "registration_month"
        new_cols.append(new_c)
    elif c == "notRepairedDamage":
        new_c = "unrepaired_damage"
        new_cols.append(new_c)
    elif c == "dateCreated":
        new_c = "ad_created"
        new_cols.append(new_c)
    else:
        new_c = change_case(c)
        new_cols.append(new_c)
print(new_cols)    
autos.columns = new_cols 
autos.head()

['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest', 'vehicle_type', 'registration_year', 'gearbox', 'power_p_s', 'model', 'odometer', 'registration_month', 'fuel_type', 'brand', 'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code', 'last_seen']


date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_p_s   model   odometer  registration_month fuel_type  \
0    manuell        158  andere  150,000km                   3       lpg   
1  automatik        286     7er  150,000km                   6    benzin   
2    manuell        102    golf   70,000km                   7    benzin   
3  automatik         71  fortwo   70,000km                   6    benzin   
4    manuell          0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

# Intital Data Exploration and Cleaning

In [5]:
autos.describe(include = 'all')

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-23 18:39:34  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox     power_p_s  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

From the prelimnary investigation of columns and their values, we can make the following observations


1. There are a number of text columns where all (or nearly all) of the values are the same like seller and offer_type
2. The column nr_of_pictures has only one value 0 
3. The columns price and odometer can to converted to numeric type though data present in string type 

In the following steps, 


1. We'll drop nr_of_pictures, seller and offer types columns as we noted as mostly one value
2. we convert the values in price and odometer columns that are stored as strings into numeric values which will help us in further analysis

In [6]:
autos = autos.drop(["nr_of_pictures", "seller", "offer_type"], axis=1)

In [7]:
autos["price"] = autos["price"].str.replace("$","").str.replace(",","").astype(int)
autos["price"].head()

0    5000
1    8500
2    8990
3    4350
4    1350
Name: price, dtype: int64

In [8]:
autos["odometer"] = autos["odometer"].str.replace(",","").str.replace("km","").astype(int)
autos["odometer"].head()

0    150000
1    150000
2     70000
3     70000
4    150000
Name: odometer, dtype: int64

In [9]:
autos = autos.rename({"odometer": "odometer_km"}, axis = 1)

# Exploring odometer and price

In [10]:
prices_unique = autos["price"].unique().shape
print(prices_unique)
price_desc = autos["price"].describe()
print(price_desc)
autos["price"].value_counts().head(20)

(2357,)
count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64


0       1421
500      781
1500     734
2500     643
1000     639
1200     639
600      531
800      498
3500     498
2000     460
999      434
750      433
900      420
650      419
850      410
700      395
4500     394
300      384
2200     382
950      379
Name: price, dtype: int64

There are 2357 unique values in the column and the prices in this column look rounded.

There are 1,421 cars listed with $0 price - given that this is only 2 percent of the of the cars, we might consider removing these rows. The maximum price is one hundred million dollars, which seems a lot, we will investigate this further.

In [11]:
price_asc = autos["price"].value_counts().sort_index().head(20)
print(price_asc)

0     1421
1      156
2        3
3        1
5        2
8        1
9        1
10       7
11       2
12       3
13       2
14       1
15       2
17       3
18       1
20       4
25       5
29       1
30       7
35       1
Name: price, dtype: int64


In [12]:
price_dsc = autos["price"].value_counts().sort_index(ascending = False).head(20)
print(price_dsc)

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
259000      1
250000      1
220000      1
198000      1
197000      1
Name: price, dtype: int64


There are a number of listings with prices below \$30, including about 1,500 at \$0. There are also a small number of listings with very high values, including 14 at around or over $1 million.

Given that eBay is an auction site, there could legitimately be items where the opening bid is \$1. We will keep the \$1 items, but remove anything above \$350,000, since it seems that prices see a sharp jump after that number and to unrealistic numbers.

In [13]:
autos = autos[autos["price"].between(1,351000)]
autos["price"].describe()

count     48565.000000
mean       5888.935591
std        9059.854754
min           1.000000
25%        1200.000000
50%        3000.000000
75%        7490.000000
max      350000.000000
Name: price, dtype: float64

In [14]:
price_asc = autos["price"].value_counts().sort_index().head(10)
price_dsc = autos["price"].value_counts().sort_index(ascending = False).head(10)
print(price_asc)
print(price_dsc)

1     156
2       3
3       1
5       2
8       1
9       1
10      7
11      2
12      3
13      2
Name: price, dtype: int64
350000    1
345000    1
299000    1
295000    1
265000    1
259000    1
250000    1
220000    1
198000    1
197000    1
Name: price, dtype: int64


In [15]:
odometer_unique = autos["odometer_km"].unique().shape
print(odometer_unique)
odometer_desc = autos["odometer_km"].describe()
print(odometer_desc)
autos["odometer_km"].value_counts()

(13,)
count     48565.000000
mean     125770.101925
std       39788.636804
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64


150000    31414
125000     5057
100000     2115
90000      1734
80000      1415
70000      1217
60000      1155
50000      1012
5000        836
40000       815
30000       780
20000       762
10000       253
Name: odometer_km, dtype: int64

We can see that the values in this field are rounded, which might indicate that sellers had to choose from pre-set options for this field. Additionally, there are more high mileage than low mileage vehicles

# Exploring the date columns

There are a number of columns with date information:

1. date_crawled
2. registration_month
3. registration_year
4. ad_created
5. last_seen

We'll explore each of these columns to learn more about the listings.

In [16]:
autos[['date_crawled','ad_created','last_seen']].head()

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

The first 10 letters in the string date_crawled correpond to only the date information without the time stamp. We look at this data to explore the dates 

In [17]:
date_cr_asc = autos["date_crawled"].str[:10].value_counts(normalize = True, dropna = False).sort_index()
print(date_cr_asc)

2016-03-05    0.025327
2016-03-06    0.014043
2016-03-07    0.036014
2016-03-08    0.033296
2016-03-09    0.033090
2016-03-10    0.032184
2016-03-11    0.032575
2016-03-12    0.036920
2016-03-13    0.015670
2016-03-14    0.036549
2016-03-15    0.034284
2016-03-16    0.029610
2016-03-17    0.031628
2016-03-18    0.012911
2016-03-19    0.034778
2016-03-20    0.037887
2016-03-21    0.037373
2016-03-22    0.032987
2016-03-23    0.032225
2016-03-24    0.029342
2016-03-25    0.031607
2016-03-26    0.032204
2016-03-27    0.031092
2016-03-28    0.034860
2016-03-29    0.034099
2016-03-30    0.033687
2016-03-31    0.031834
2016-04-01    0.033687
2016-04-02    0.035478
2016-04-03    0.038608
2016-04-04    0.036487
2016-04-05    0.013096
2016-04-06    0.003171
2016-04-07    0.001400
Name: date_crawled, dtype: float64


By looking at the date range analysis of date_crawled, we can see that all the ads are crawled between 2016-03-05 and 2016-04-07 that is in a span of roughly one month, most of them in March of 2016. And the maximum number of ads crawled on 3rd of April 2016

In [18]:
ad_cr_unique = autos["ad_created"].str[:10].unique().shape
print(ad_cr_unique)
ad_cr_asc = autos["ad_created"].str[:10].value_counts(normalize = True, dropna = False).sort_index()
print(ad_cr_asc)

(76,)
2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
2015-12-30    0.000021
2016-01-03    0.000021
2016-01-07    0.000021
2016-01-10    0.000041
2016-01-13    0.000021
2016-01-14    0.000021
2016-01-16    0.000021
2016-01-22    0.000021
2016-01-27    0.000062
2016-01-29    0.000021
2016-02-01    0.000021
2016-02-02    0.000041
2016-02-05    0.000041
2016-02-07    0.000021
2016-02-08    0.000021
2016-02-09    0.000021
2016-02-11    0.000021
2016-02-12    0.000041
2016-02-14    0.000041
2016-02-16    0.000021
2016-02-17    0.000021
2016-02-18    0.000041
2016-02-19    0.000062
2016-02-20    0.000041
2016-02-21    0.000062
                ...   
2016-03-09    0.033151
2016-03-10    0.031895
2016-03-11    0.032904
2016-03-12    0.036755
2016-03-13    0.017008
2016-03-14    0.035190
2016-03-15    0.034016
2016-03-16    0.030125
2016-03-17    0.031278
2016-03-18    0.013590
2016-03-19    0.033687
2016-03-20    0.037949
2016-

By looking at the date range analysis of ad_created column of autos dataframe, we can see that the ads enlisted on eBay each day before 2016-03-09 account for a negligble percentage of ads enlisted in comparison to ads enlisted later. 
There is a large variety of ad created dates. Most fall within 1-2 months of the listing date, but a few are quite old, with the oldest at around 9 months. The last date on which the ads are enlisted is the same as the last last date on which ads are crawled. And the most of the ads are enlisted b/w March and April,2016 

In [19]:
last_seen_unique = autos["last_seen"].str[:10].unique().shape
print(last_seen_unique)
last_seen_asc = autos["last_seen"].str[:10].value_counts(normalize = True, dropna = False).sort_index()
print(last_seen_asc)

(34,)
2016-03-05    0.001071
2016-03-06    0.004324
2016-03-07    0.005395
2016-03-08    0.007413
2016-03-09    0.009595
2016-03-10    0.010666
2016-03-11    0.012375
2016-03-12    0.023783
2016-03-13    0.008895
2016-03-14    0.012602
2016-03-15    0.015876
2016-03-16    0.016452
2016-03-17    0.028086
2016-03-18    0.007351
2016-03-19    0.015834
2016-03-20    0.020653
2016-03-21    0.020632
2016-03-22    0.021373
2016-03-23    0.018532
2016-03-24    0.019767
2016-03-25    0.019211
2016-03-26    0.016802
2016-03-27    0.015649
2016-03-28    0.020859
2016-03-29    0.022341
2016-03-30    0.024771
2016-03-31    0.023783
2016-04-01    0.022794
2016-04-02    0.024915
2016-04-03    0.025203
2016-04-04    0.024483
2016-04-05    0.124761
2016-04-06    0.221806
2016-04-07    0.131947
Name: last_seen, dtype: float64


The crawler recorded the date it last saw any listing, which allows us to determine on what day a listing was removed, presumably because the car was sold.

The last three days contain a disproportionate amount of 'last seen' values. Given that these are 6-10x the values from the previous days, it's unlikely that there was a massive spike in sales, and more likely that these values are to do with the crawling period ending and don't indicate car sales

In [20]:
reg_year_desc = autos["registration_year"].describe()
reg_year_unique = autos["registration_year"].unique().shape
reg_year_asc = autos["registration_year"].value_counts(normalize = True, dropna = False).sort_index().head(20)
reg_year_dsc = autos["registration_year"].value_counts(normalize = True, dropna = False).sort_index(ascending = False).head(20)
print(reg_year_desc)
print(reg_year_unique)
print(reg_year_asc)
print(reg_year_dsc)

count    48565.000000
mean      2004.755421
std         88.643887
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64
(95,)
1000    0.000021
1001    0.000021
1111    0.000021
1800    0.000041
1910    0.000103
1927    0.000021
1929    0.000021
1931    0.000021
1934    0.000041
1937    0.000082
1938    0.000021
1939    0.000021
1941    0.000041
1943    0.000021
1948    0.000021
1950    0.000062
1951    0.000041
1952    0.000021
1953    0.000021
1954    0.000041
Name: registration_year, dtype: float64
9999    0.000062
9000    0.000021
8888    0.000021
6200    0.000021
5911    0.000021
5000    0.000082
4800    0.000021
4500    0.000021
4100    0.000021
2800    0.000021
2019    0.000041
2018    0.009678
2017    0.028663
2016    0.025121
2015    0.008072
2014    0.013652
2013    0.016535
2012    0.026974
2011    0.033419
2010    0.032719
Name: registration_year, dtype: float64


By investigating the registration year column, we find out that there are vehicles registered in years that do not make sense like before 1800 i.e, before cars are invented  and 2016 and after i.e after which the car was first listed on eBay.The number of cars registed before 1900 and after 2016 are around 2000 which account for around 5 percentage of total cars enlisted. 
Hence, we consider 2016 as the highest acceptable year and 1900 as the lowest acceptable year to clean the data 

In [21]:
autos = autos[(autos["registration_year"]>1900)&(autos["registration_year"]<2017)]
autos["registration_year"].value_counts(normalize = True, dropna = False).head(20)

2000    0.067608
2005    0.062895
1999    0.062060
2004    0.057904
2003    0.057818
2006    0.057197
2001    0.056468
2002    0.053255
1998    0.050620
2007    0.048778
2008    0.047450
2009    0.044665
1997    0.041794
2011    0.034768
2010    0.034040
1996    0.029412
2012    0.028063
1995    0.026285
2016    0.026135
2013    0.017202
Name: registration_year, dtype: float64

It appears that most of the vehicles were first registered in the past 20 years.

# Exploring price by brand

In [22]:
autos["brand"].value_counts(normalize = True)

volkswagen        0.211264
bmw               0.110045
opel              0.107581
mercedes_benz     0.096463
audi              0.086566
ford              0.069900
renault           0.047150
peugeot           0.029841
fiat              0.025642
seat              0.018273
skoda             0.016409
nissan            0.015274
mazda             0.015188
smart             0.014160
citroen           0.014010
toyota            0.012703
hyundai           0.010025
sonstige_autos    0.009811
volvo             0.009147
mini              0.008762
mitsubishi        0.008226
honda             0.007840
kia               0.007069
alfa_romeo        0.006641
porsche           0.006127
suzuki            0.005934
chevrolet         0.005698
chrysler          0.003513
dacia             0.002635
daihatsu          0.002506
jeep              0.002271
subaru            0.002142
land_rover        0.002099
saab              0.001649
jaguar            0.001564
daewoo            0.001500
trabant           0.001392
r

German manufacturers represent four out of the top five brands, accounting for almost 50% of the overall listings. Volkswagen is by far the most popular brand. 

There are lots of brands that don't have a significant percentage of listings, so we will limit our analysis to brands representing more than 5% of total listings

In [23]:
brand_vc = autos["brand"].value_counts(normalize = True)
common_brands = brand_vc[brand_vc > 0.05].index
print(common_brands)

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford'], dtype='object')


In [24]:
brand_mean_price = {}
for b in common_brands:
    selected_rows = autos.loc[autos["brand"] == b,"price"]
    m = selected_rows.mean()
    brand_mean_price[b]= int(m)
print(brand_mean_price)

{'bmw': 8332, 'volkswagen': 5402, 'audi': 9336, 'opel': 2975, 'ford': 3749, 'mercedes_benz': 8628}


There are 40 unique brands of cars in the given database with only 6 brands having cars enlisted over 5 percentage (2326)of total cars enlisted on eBay site. Above, we aggregated the data of each unique brand with number of cars enlisted greater than 5 percentage of total cars and calculated the mean price of aggreagted data over that chosen unique brand.
The results show that Audi brand has the highest mean price of all the other brands enlisted on the site
And renault has the lowest mean price of all brands 
The price gap between the top 3 and bottom 3 brands is distinctive

In [25]:
brand_mean_milage = {}
for b in common_brands:
    selected_rows = autos.loc[autos["brand"] == b,"odometer_km"]
    m = selected_rows.mean()
    brand_mean_milage[b]= int(m)
print(brand_mean_milage)


{'bmw': 132572, 'volkswagen': 128707, 'audi': 129157, 'opel': 129310, 'ford': 124266, 'mercedes_benz': 130788}


In [26]:
bmp_series = pd.Series(brand_mean_price).sort_values(ascending = False)
bmm_series = pd.Series(brand_mean_milage).sort_values(ascending = False)
print(bmp_series)
print(bmm_series)
brand_info = pd.DataFrame(bmp_series,columns=["mean_price"])
brand_info["mean_milage"] = bmm_series
brand_info

audi             9336
mercedes_benz    8628
bmw              8332
volkswagen       5402
ford             3749
opel             2975
dtype: int64
bmw              132572
mercedes_benz    130788
opel             129310
audi             129157
volkswagen       128707
ford             124266
dtype: int64


mean_price  mean_milage
audi                 9336       129157
mercedes_benz        8628       130788
bmw                  8332       132572
volkswagen           5402       128707
ford                 3749       124266
opel                 2975       129310

Above, we aggregated the data of each unique brand with number of cars enlisted greater than 5% of total cars and calculated the mean milage of aggreagted data over that chosen unique brand. The results show that 


1. bmw brand has the highest mean milage of all the other brands enlisted on the site and opel has the lowest mean price of all brands 
2. The gap between mean milage of the top 3 and bottom 3 brands is not much and all of them lie in a narrow range of mean milage. 3. The difference between mean milages between top 2 brands and bottom 2 brands is not as distinctive as the prices

# Finding most common brand/model combinations

In [29]:
brand_model = {}
for b in autos["brand"].unique():
    model_count = autos.loc[autos["brand"] == b,"model"].value_counts(dropna = False)
    model_unique = model_count.index[0]
    brand_model[b] = model_unique
bm_series = pd.Series(brand_model)
bm_series

alfa_romeo               156
audi                      a4
bmw                      3er
chevrolet             andere
chrysler              andere
citroen               andere
dacia                sandero
daewoo                 matiz
daihatsu               cuore
fiat                   punto
ford                   focus
honda                  civic
hyundai              i_reihe
jaguar                andere
jeep                   grand
kia                   andere
lada                    niva
lancia               ypsilon
land_rover        freelander
mazda                3_reihe
mercedes_benz       c_klasse
mini                  cooper
mitsubishi              colt
nissan                 micra
opel                   corsa
peugeot              2_reihe
porsche                  911
renault               twingo
rover                 andere
saab                  andere
seat                   ibiza
skoda                octavia
smart                 fortwo
sonstige_autos           NaN
subaru        

# Exploring average prices among different milage groups

In [30]:
autos["odometer_km"].value_counts()

150000    30085
125000     4857
100000     2058
90000      1673
80000      1375
70000      1187
60000      1128
50000       993
40000       797
5000        785
30000       760
20000       742
10000       241
Name: odometer_km, dtype: int64

In [31]:
mean_prices = {}
for m in autos["odometer_km"].unique():
    avg_price = autos.loc[autos["odometer_km"]==m,"price"].mean()
    mean_prices[m] = avg_price
mp_series = pd.Series(mean_prices)
mp_series
    

5000       8873.515924
10000     20550.867220
20000     18448.477089
30000     16608.836842
40000     15499.568381
50000     13812.173212
60000     12385.004433
70000     10927.182814
80000      9721.947636
90000      8465.025105
100000     8132.697279
125000     6214.022030
150000     3767.927107
dtype: float64

1. Average prices are inversely related to milage. That is, higher milage groups have lower prices and vice versa
2. Except, one outlier that is the milage group 5000 km whose average price lies between the milage groups of 80000 km and 90000 km

## Comparing prices of cars with damnage vs non-damaged cars

In [32]:
autos["unrepaired_damage"].value_counts(dropna = False)

nein    33834
NaN      8307
ja       4540
Name: unrepaired_damage, dtype: int64

In [33]:
yes_no = {"nein" : "No", "ja" : "Yes"}
autos["unrepaired_damage"] = autos["unrepaired_damage"].map(yes_no)
autos["unrepaired_damage"].value_counts(dropna = False)

No     33834
NaN     8307
Yes     4540
Name: unrepaired_damage, dtype: int64

In [34]:
d_prices = {}
for d in autos["unrepaired_damage"].unique():
    avg_price = autos.loc[autos["unrepaired_damage"]==d,"price"].mean()
    d_prices[d] = avg_price
dp_series = pd.Series(d_prices)
dp_series

NaN            NaN
Yes    2241.146035
No     7164.033103
dtype: float64

There appears to be a significant price gap between cars with damage and non-damaged cars as can be seen from the average prices. There exists a gap of around \$5000 

## Converting german wrods to english 

In [35]:
autos.columns

Index(['date_crawled', 'name', 'price', 'abtest', 'vehicle_type',
       'registration_year', 'gearbox', 'power_p_s', 'model', 'odometer_km',
       'registration_month', 'fuel_type', 'brand', 'unrepaired_damage',
       'ad_created', 'postal_code', 'last_seen'],
      dtype='object')

In [36]:
autos["gearbox"].unique()

array(['manuell', 'automatik', nan], dtype=object)

In [37]:
gb = {"manuell" : "Manual", "automatik" : "Automatic"}
autos["gearbox"] = autos["gearbox"].map(gb)
autos["gearbox"].value_counts(dropna = False)

Manual       34715
Automatic     9856
NaN           2110
Name: gearbox, dtype: int64

In [38]:
autos["fuel_type"].unique()

array(['lpg', 'benzin', 'diesel', nan, 'cng', 'hybrid', 'elektro',
       'andere'], dtype=object)

In [40]:
ft = {"benzin" : "benzine", "lpg" : "lpg", "diesel" :"diesel", "cng" : "cng", "hybrid" :"hybrid", "elektro" : "electro", "andere" : "other"}
autos["fuel_type"] = autos["fuel_type"].map(ft)
autos["fuel_type"].value_counts(dropna = False)

NaN    46681
Name: fuel_type, dtype: int64